### PR TITLE
Customer Ticket 1 

### PR DESCRIPTION
**Pull request by Jengwen**
## Proposed changes
This is the build to run the Customer Controller and it's associated tests. You can now Get a list of All Customers, Get a single customer, Add a customer, and edit a customer.
You can also query to include a list of all customers with their payment types, or query to include a list of customers with their list of products for sale, or query for a customer's last name or first name with a string of included letters.  You can also test customers Get All method, Get a single, post and put methods.

## Steps To Test
Run the following SQL Script:
DELETE FROM OrderProduct;
DELETE FROM ComputerEmployee;
DELETE FROM EmployeeTraining;
DELETE FROM Employee;
DELETE FROM TrainingProgram;
DELETE FROM Computer;
DELETE FROM Department;
DELETE FROM [Order];
DELETE FROM PaymentType;
DELETE FROM Product;
DELETE FROM ProductType;
DELETE FROM Customer;
ALTER TABLE Employee DROP CONSTRAINT [FK_EmployeeDepartment];
ALTER TABLE ComputerEmployee DROP CONSTRAINT [FK_ComputerEmployee_Employee];
ALTER TABLE ComputerEmployee DROP CONSTRAINT [FK_ComputerEmployee_Computer];
ALTER TABLE EmployeeTraining DROP CONSTRAINT [FK_EmployeeTraining_Employee];
ALTER TABLE EmployeeTraining DROP CONSTRAINT [FK_EmployeeTraining_Training];
ALTER TABLE Product DROP CONSTRAINT [FK_Product_ProductType];
ALTER TABLE Product DROP CONSTRAINT [FK_Product_Customer];
ALTER TABLE PaymentType DROP CONSTRAINT [FK_PaymentType_Customer];
ALTER TABLE [Order] DROP CONSTRAINT [FK_Order_Customer];
ALTER TABLE [Order] DROP CONSTRAINT [FK_Order_Payment];
ALTER TABLE OrderProduct DROP CONSTRAINT [FK_OrderProduct_Product];
ALTER TABLE OrderProduct DROP CONSTRAINT [FK_OrderProduct_Order];

DROP TABLE IF EXISTS OrderProduct;
DROP TABLE IF EXISTS ComputerEmployee;
DROP TABLE IF EXISTS EmployeeTraining;
DROP TABLE IF EXISTS Employee;
DROP TABLE IF EXISTS TrainingProgram;
DROP TABLE IF EXISTS Computer;
DROP TABLE IF EXISTS Department;
DROP TABLE IF EXISTS [Order];
DROP TABLE IF EXISTS PaymentType;
DROP TABLE IF EXISTS Product;
DROP TABLE IF EXISTS ProductType;
DROP TABLE IF EXISTS Customer;

CREATE TABLE Department (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	[Name] VARCHAR(55) NOT NULL,
	Budget 	INTEGER NOT NULL
);

CREATE TABLE Employee (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	FirstName VARCHAR(55) NOT NULL,
	LastName VARCHAR(55) NOT NULL,
	DepartmentId INTEGER NOT NULL,
	IsSuperVisor BIT NOT NULL DEFAULT(0),
    CONSTRAINT FK_EmployeeDepartment FOREIGN KEY(DepartmentId) REFERENCES Department(Id)
);

CREATE TABLE Computer (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	PurchaseDate DATETIME NOT NULL,
	DecomissionDate DATETIME,
	Make VARCHAR(55) NOT NULL,
	Manufacturer VARCHAR(55) NOT NULL
);

CREATE TABLE ComputerEmployee (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	EmployeeId INTEGER NOT NULL,
	ComputerId INTEGER NOT NULL,
	AssignDate DATETIME NOT NULL,
	UnassignDate DATETIME,
    CONSTRAINT FK_ComputerEmployee_Employee FOREIGN KEY(EmployeeId) REFERENCES Employee(Id),
    CONSTRAINT FK_ComputerEmployee_Computer FOREIGN KEY(ComputerId) REFERENCES Computer(Id)
);

CREATE TABLE TrainingProgram (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	[Name] VARCHAR(255) NOT NULL,
	StartDate DATETIME NOT NULL,
	EndDate DATETIME NOT NULL,
	MaxAttendees INTEGER NOT NULL
);

CREATE TABLE EmployeeTraining (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	EmployeeId INTEGER NOT NULL,
	TrainingProgramId INTEGER NOT NULL,
    CONSTRAINT FK_EmployeeTraining_Employee FOREIGN KEY(EmployeeId) REFERENCES Employee(Id),
    CONSTRAINT FK_EmployeeTraining_Training FOREIGN KEY(TrainingProgramId) REFERENCES TrainingProgram(Id)
);

CREATE TABLE ProductType (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	[Name] VARCHAR(55) NOT NULL
);

CREATE TABLE Customer (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	FirstName VARCHAR(55) NOT NULL,
	LastName VARCHAR(55) NOT NULL
);

CREATE TABLE Product (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	ProductTypeId INTEGER NOT NULL,
	CustomerId INTEGER NOT NULL,
	Price MONEY NOT NULL,
	Title VARCHAR(255) NOT NULL,
	[Description] VARCHAR(255) NOT NULL,
	Quantity INTEGER NOT NULL,
    CONSTRAINT FK_Product_ProductType FOREIGN KEY(ProductTypeId) REFERENCES ProductType(Id),
    CONSTRAINT FK_Product_Customer FOREIGN KEY(CustomerId) REFERENCES Customer(Id)
);

CREATE TABLE PaymentType (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	AcctNumber INTEGER NOT NULL,
	[Name] VARCHAR(55) NOT NULL,
	CustomerId INTEGER NOT NULL,
    CONSTRAINT FK_PaymentType_Customer FOREIGN KEY(CustomerId) REFERENCES Customer(Id)
);

CREATE TABLE [Order] (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	CustomerId INTEGER NOT NULL,
	PaymentTypeId INTEGER,
    CONSTRAINT FK_Order_Customer FOREIGN KEY(CustomerId) REFERENCES Customer(Id),
    CONSTRAINT FK_Order_Payment FOREIGN KEY(PaymentTypeId) REFERENCES PaymentType(Id)
);

CREATE TABLE OrderProduct (
	Id INTEGER NOT NULL PRIMARY KEY IDENTITY,
	OrderId INTEGER NOT NULL,
	ProductId INTEGER NOT NULL,
    CONSTRAINT FK_OrderProduct_Product FOREIGN KEY(ProductId) REFERENCES Product(Id),
    CONSTRAINT FK_OrderProduct_Order FOREIGN KEY(OrderId) REFERENCES [Order](Id)
);
INSERT INTO Computer (PurchaseDate, Make, Manufacturer) VALUES ('06/26/1982','Latitude', 'Dell')
INSERT INTO Computer (PurchaseDate, Make, Manufacturer) VALUES ('04/16/1999','Inspiron', 'Dell')
INSERT INTO Computer (PurchaseDate, Make, Manufacturer) VALUES ('03/13/2010','Latitude', 'Dell')
INSERT INTO Customer (FirstName, LastName) VALUES ('Fred','Smith')
INSERT INTO Customer (FirstName, LastName) VALUES ('Sally','Jones')
INSERT INTO Customer (FirstName, LastName) VALUES ('Bob','Robertson')
INSERT INTO Department (Name, Budget) VALUES ('Sales', 501245)
INSERT INTO Department (Name, Budget) VALUES ('Marketing', 99801)
INSERT INTO Department (Name, Budget) VALUES ('Human Resources', 50879)
INSERT INTO ProductType (Name) VALUES ('Kitchen Wares')
INSERT INTO ProductType (Name) VALUES ('Toys')
INSERT INTO ProductType (Name) VALUES ('Shoes')
INSERT INTO TrainingProgram (Name, StartDate, EndDate, MaxAttendees) VALUES ('New Hire Orientation', '02/13/2011', '03/02/2019', 50)
INSERT INTO TrainingProgram (Name, StartDate, EndDate, MaxAttendees) VALUES ('New Hire Orientation', '05/13/2018', '01/22/2019', 50)
INSERT INTO TrainingProgram (Name, StartDate, EndDate, MaxAttendees) VALUES ('New Hire Orientation', '08/13/2001', '06/02/2015', 50)
INSERT INTO Employee (FirstName, LastName, DepartmentId, IsSupervisor) VALUES ('Fred','Smith', 2, 0)
INSERT INTO Employee (FirstName, LastName, DepartmentId, IsSupervisor) VALUES ('John','Johnson', 2, 0)
INSERT INTO Employee (FirstName, LastName, DepartmentId, IsSupervisor) VALUES ('Larry','Larryson', 3, 1)
INSERT INTO EmployeeTraining (EmployeeId, TrainingProgramId) VALUES (2, 1)
INSERT INTO EmployeeTraining (EmployeeId, TrainingProgramId) VALUES (1, 1)
INSERT INTO EmployeeTraining (EmployeeId, TrainingProgramId) VALUES (2, 2)
INSERT INTO PaymentType (AcctNumber, Name, CustomerId) VALUES (9001,'Steve Stevenson', 1)
INSERT INTO PaymentType (AcctNumber, Name, CustomerId) VALUES (9002, 'Mary Maryson', 1)
INSERT INTO PaymentType (AcctNumber, Name, CustomerId) VALUES (9003, 'Mike Michaelson', 2)
INSERT INTO Product (ProductTypeId, CustomerId, Price, Title, Description, Quantity) VALUES (2, 2, 9.99, 'Nerf Gun', 'Big gun shoots soft bullets', 1)
INSERT INTO Product (ProductTypeId, CustomerId, Price, Title, Description, Quantity) VALUES (1, 2, 3.99, 'Dawn Dish Soap', 'Takes grease out of the way', 1)
INSERT INTO Product (ProductTypeId, CustomerId, Price, Title, Description, Quantity) VALUES (3, 1, 6.99, 'Sandals', 'Beach Shoes', 1)
INSERT INTO [Order] (CustomerId, PaymentTypeId) VALUES (2, 1)
INSERT INTO [Order] (CustomerId, PaymentTypeId) VALUES (1, 2)
INSERT INTO [Order] (CustomerId, PaymentTypeId) VALUES (1, 1)

To get a list of customers:
Start the application, in the browser type "http://localhost:63540/api/customer" you should see a list of customers
To get a single customer:
in the browser type "http://localhost:63540/api/customer/1"
you should see a customer by the name of Fred Smith
To query to include a list of payment types:
in the browser type "http://localhost:63540/api/customer?include=payments"
You should see Sally Jones and Bob Robertson with a list of their payment types
To query to include a list of products for sale by a customer
in the browser type "http://localhost:63540/api/customer?include=products"
You will then see Sally Jones with a list of her products for sale as well as Bob Robertson with his products for sale 
To test POST Method:
Type http://localhost:63540/api/customer and change dropdown to POST in Postmaster and the following into the body in Postman 
{
"firstName": "Judy",
"lastName": "Dench"
} 
hit send and get all customers again with http://localhost:63540/api/customer in the browser
you should now see your new customer in the customer list
To test PUT Method
Type http://localhost:63540/api/customer/4 and change dropdown to PUT in Postmaster and the following into the body in Postman 
{
"firstName": "Jane",
"lastName": Dench"
}
 hit send and get all customers again with http://localhost:63540/api/customer in the browser
you should now see your edited customer name in the list

You can also select Test and Run all Tests in Visual Studio to see successful tests

## What kind of change does this PR introduce?
Please check one of the following:
- [ ] Bug fix
- [ ] Documentation update
- [ x] New feature
- [ ] Other

### CODE REVIEW

**Comment by LanceP51** [comment]
need to go to Appsettings and remove "01" from SQLServer address.
need to use customer/4 when trying out the PUT method

works well, without issues
Jen will be changing this SQL setting so we wont have to worry with this in the future

**Review by Tommy-Spurlock — CHANGES_REQUESTED** [review]
Please include SQL script data needed to run test.

**Comment by Jengwen** [comment]
The SQL script is now in the testing instructions

**Review by Tommy-Spurlock — CHANGES_REQUESTED** [review]
When I tried to run the SQL script it threw errors concerning foreign key constraints. Check to see if your SQL script needs any additional information. 
*When running the get all at `/api/customer` I received 7 items.
* `include=products` threw an error. 
* When running `PUT` in postman I received a 400 error. 

I believe these are all errors related to the SQL script seed data. You may need to include a SQL command that will recreate the database and tables before seeding data.

**Comment by Jengwen** [comment]
Updated the SQL script to correct table issues.

File: BangazonAPITest/CustomerControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                HttpResponseMessage response = await client.GetAsync(url);
+                response.EnsureSuccessStatusCode();
+
+                // Convert to JSON
+                string responseBody = await response.Content.ReadAsStringAsync();
+
+                // Convert from JSON to C#
+                List<Customer> customers = JsonConvert.DeserializeObject<List<Customer>>(responseBody);
+
+                // Make sure we got back a 200 OK Status and that there are more than 0 coffees in our database
```
Comment (LanceP51) [inline BangazonAPITest/CustomerControllerTests.cs]: coffee?

File: BangazonAPITest/CustomerControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        }
+        // test to get all customers
+        [Fact]
+        public async Task Get_All_Customer()
+        {
+
+            using (var client = new APIClientProvider().Client)
+            {
+
+                // Try to get all of the students from /api/customer
```
Comment (LanceP51) [inline BangazonAPITest/CustomerControllerTests.cs]: student?

File: BangazonAPITest/CustomerControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+                // Store the JSON body of the response
+                string responseBody = await response.Content.ReadAsStringAsync();
+
+                // Deserialize the JSON into an instance of Student
+                Customer newlyCreatedCustomer = JsonConvert.DeserializeObject<Customer>(responseBody);
+
+                return newlyCreatedCustomer;
+            }
+        }
+        // Reusable method to deelte a customer from the database
```
Comment (LanceP51) [inline BangazonAPITest/CustomerControllerTests.cs]: deelte?

File: BangazonAPITest/CustomerControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        {
+            firstName = "Joe",
+            lastName = "Latte"
+          
+        };
+
+        // We'll store our base url for this route as a private field to avoid typos
+        private string url { get; } = "/api/customer";
+
+        // Reusable method to create a new student in the database and return it
```
Comment (LanceP51) [inline BangazonAPITest/CustomerControllerTests.cs:26]: student? or customer?

File: BangazonAPITest/CustomerControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                Assert.Equal(dummyCustomer.firstName, JoeCustomerFromDB.firstName);
+                Assert.Equal(dummyCustomer.lastName, JoeCustomerFromDB.lastName);
+
+                // Clean up after ourselves-- delete the dummy customer we just created
+                await deleteDummyCustomer(JoeCustomerFromDB);
+
+            }
+        }
+        [Fact]
+        public async Task Update_Student()
```
Comment (LanceP51) [inline BangazonAPITest/CustomerControllerTests.cs]: student?

File: BangazonAPITest/CustomerControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        {
+
+            using (var client = new APIClientProvider().Client)
+            {
+                // Create a dummy customer
+                Customer newJoeCustomer = await CreateDummyCustomer();
+
+                // Make a new title and assign it to our dummy student
```
Comment (LanceP51) [inline BangazonAPITest/CustomerControllerTests.cs]: student?

**Review by LanceP51 — COMMENTED** [review]
please correct typos

**Review by Tommy-Spurlock — CHANGES_REQUESTED** [review]
Include=products returned both sally jones and Fred Smith twice. The testing instructions for`PUT` only include firstName and should also include lastName 
`{
"firstName": "Jane",
"lastName": "Dench"
}`.    Almost There!!!

**Comment by AllxnCook** [comment]
functions on my machine
saw no errors in the process of testing
all tests passed
great job!

**Pull request merged**